tt_metal/impl/buffers/global_semaphore.hpp

```diff
@@ -0,0 +1,42 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <cstdint>
+#include <memory>
+
+#include "tt_metal/common/core_coord.hpp"
+#include "tt_metal/impl/buffers/buffer_constants.hpp"
+#include "tt_metal/llrt/hal.hpp"
+
+namespace tt::tt_metal {
+
+inline namespace v0 {
+class Buffer;
+class Device;
+}  // namespace v0
+
+class GlobalSemaphore {
+   public:
+    GlobalSemaphore(Device *device, CoreRangeSet cores, uint32_t initial_value, BufferType buffer_type = BufferType::L1);
```

Comment:
Probably want to take `CoreRangeSet` as `const &` or `&&`, or possibly overloads for both, but at best this forces an extra move-construction more than necessary of a `std::vector` if users explicitly `std::move(cores)` to this parameter or construct it in-place, and at worst it forces an extra copy-construction if they don't.

Reply:
Ah whoops, meant to change it to `const &` before putting up the pr. Will add the move constructor as well.